### ISSUE TITLE
Threema Gateway support

### ISSUE THREAD

**Issue by rugk** [nextcloud/twofactor_sms#46]
It would be great if you could – in addition to Telegram – also support Threema.

In contrast to all other methods here (AFAIK), it is at least end-to-end-encrypted. :smiley: 

For more information, see https://gateway.threema.ch/.

**Comment by ChristophWurst** [comment]
Pull requests for a Threema integration are very welcome. I don't use it right now and I'd also favor FLOSS over a proprietary messenger.

> For more information, see https://gateway.threema.ch/.

Such gateways are great, but usually messages get encrypted at the gateways, allowing it to read all the sensitive information. That's not end-to-end encrypted :wink:

**Comment by rugk** [comment]
Threema offers both. Their SDK actually does real e2e crpyto. In contrast to Telegram, BTW, but you can forget their crypto anyway. But let's not get off-topic here… :smiley:

**Comment by simonspa** [comment]
Looking at the Telegram gateway implementation here, and comparing with the SDK snippets on the Threema Gateway website, this looks fairly simple (compare section "Send a Text Message to a Threema-ID (End-To-End Mode)" on the [PHP SDK page](https://gateway.threema.ch/de/developer/sdk-php))

However, I don't fully understand the mechanisms of retrieving information about the user. Would the ID (and thus the public key) be stored in the user profile somehow? How could this be retrieved then?

**Comment by rugk** [comment]
> Would the ID (and thus the public key) be stored in the user profile somehow?

Of course, the user somehow has to provide the Threema ID. This may be done when settingupthe 2fa method, like with TOTP, or, as you said, in the user's profile.

**Comment by rugk** [comment]
BTW the hardest part may be just that libsodium is required for the server admin to install. :smile:

**Comment by WhiteBahamut** [comment]
Not sure if the Threema GW needs to be supported, but Threema just described their approach to multi-device capabilities. Maybe Talk can "just" be another Threema Device :)
https://threema.ch/en/blog
Use Threema quite a lot and it would be awsome to have it integrated into Talk, so I can mainly use Talk

**Comment by rugk** [comment]
@WhiteBahamut Mind in which repo you are commenting. This repo is https://github.com/nextcloud/twofactor_gateway, i.e. this is about integrating Threema as an 2FA provider – not as anything related to "Talk".
